### PR TITLE
fix(ttnn): eliminate 16 dead dispatches and discarded where() guards in div_bw (fixes #55392)

### LINKED ISSUE #55392: ttnn.div_bw (tensor, tensor) computes both zero-divisor where() guards and discards them: 16 of 23 dispatches are dead, and wiring the guard in would break other == -0.0
## Summary

The tensor–tensor overload of `ttnn.div_bw` builds a zero-divisor guard for each gradient with `ttnn::where(...)` and then **discards it**: the call omits the `output` argument and its return value is never assigned, so `result[0]` / `result[1]` still hold the pre-guard products.

- `grad_a` guard: [`binary_backward.cpp#L727-L735`](https://github.com/tenstorrent/tt-metal/blob/7b5ae30865b0bf771da32696c9bfcc54ec412a49/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L727-L735)
- `other_grad` guard: [`#L749-L766`](https://github.com/tenstorrent/tt-metal/blob/7b5ae30865b0bf771da32696c9bfcc54ec412a49/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L749-L766)

`ttnn::where` takes the destination as its **fifth** parameter ([`ternary.hpp#L17-L23`](https://github.com/tenstorrent/tt-metal/blob/7b5ae30865b0bf771da32696c9bfcc54ec412a49/ttnn/cpp/ttnn/operations/eltwise/ternary/ternary.hpp#L17-L23)). Both calls stop at the fourth (`output_mem_config`). Every sibling passes it: the **scalar** overload of `div_bw` two functions up ([`#L684-L689`](https://github.com/tenstorrent/tt-metal/blob/7b5ae30865b0bf771da32696c9bfcc54ec412a49/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L684-L689)), `sqrt_bw`, `rsqrt_bw`, `where_bw`.

```cpp
ttnn::multiply(grad_tensor, ttnn::reciprocal(other, ...), std::nullopt, output_mem_config, input_grad);  // writes input_grad
ttnn::where(                                          // <-- return value dropped, no output argument
    ttnn::eqz(other, output_mem_config),
    ttnn::where(ttnn::eqz(grad_tensor, ...), t_nan,
                ttnn::multiply(ttnn::sign(grad_tensor, ...), t_inf, ...), ...),
    input_grad.value(),
    output_mem_config);
result[0] = input_grad;                               // still the pre-guard value
```

### 1. Sixteen of twenty-three dispatches are dead

With both gradients requested and `rounding_mode=None`:

| block | device ops |
|---|---:|
| useful, `grad_a` (`reciprocal`, `multiply`) | 2 |
| useful, `other_grad` (`square`, `reciprocal`, `multiply`, `neg`, `multiply`) | 5 |
| **discarded** `grad_a` guard (`eqz`×2, `sign`, `multiply`, `where`×2) | **6** |
| **discarded** `other_grad` guard (`eqz`×3, `sign`×2, `multiply`×2, `where`×3) | **10** |

~70% of the op's launches allocate a fresh tensor whose contents are never read. Same class as #54825 (`polygamma_bw`, 55%) and #54827 (`logit_bw`, 23%), larger than either.

### 2. The naive fix would introduce a sign bug — the guard should be deleted, not wired in

The op currently returns the right answer *because* the guard is dropped. Working through the raw path with IEEE fp32 primitives (`reciprocal(+0)=+inf`, `reciprocal(-0)=-inf`, `square(±0)=+0`):

| `grad` | `a` | `other` | raw `dA` | guard `dA` | torch `dA` | raw `dB` | guard `dB` | torch `dB` |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | 2 | `+0.0` | inf | inf | inf | -inf | -inf | -inf |
| -1 | 2 | `+0.0` | -inf | -inf | -inf | inf | inf | inf |
| 0 | 2 | `+0.0` | nan | nan | nan | nan | nan | nan |
| 1 | 0 | `+0.0` | inf | inf | inf | nan | nan | nan |
| 1 | 2 | **`-0.0`** | **-inf** | inf | **-inf** | -inf | -inf | -inf |
| -1 | 2 | **`-0.0`** | **inf** | -inf | **inf** | inf | inf | inf |
| 1 | -3 | **`-0.0`** | **-inf** | inf | **-inf** | inf | inf | inf |

At `other == +0` the raw product already equals what the guard would write, bit for bit (`sign(grad)*inf`, `NaN` when `grad==0`). At `other == -0.0` the raw path yields `-sign(grad)*inf`, which is what autograd returns; the guard hard-codes `+sign(grad)*inf` and would be **wrong**. So adding the missing `input_grad` / `other_grad` argument — the obvious one-line "fix" — would regress negative-zero divisors.

## Reproducer

Host only, no device required (`pip install torch numpy`):

<details><summary>repro_div_bw_guard.py</summary>

```python
"""Host-only model of ttnn.div_bw (tensor, tensor), tt-metal @ 7b5ae30,
binary_backward.cpp:719-768.

Shows that (a) the two zero-divisor ttnn::where guards are computed and
discarded (no output argument, return value unused), (b) the raw
grad * reciprocal(other) path already produces the guard's values at
other == +0, and (c) at other == -0.0 the raw path matches torch while the
discarded guard would NOT have. No device required.
Assumes IEEE fp32 primitives: reciprocal(+0)=+inf, reciprocal(-0)=-inf, sign(0)=0.
"""
import numpy as np
import torch

f32 = np.float32
INF, NAN = f32(np.inf), f32(np.nan)


def raw_grad_a(grad, other):
    with np.errstate(divide="ignore", invalid="ignore"):
        return (grad * (f32(1) / other)).astype(f32)


def guard_grad_a(grad, other, raw):
    """What the discarded where() at :727 would have written."""
    g = np.where(grad == 0, NAN, np.sign(grad) * INF).astype(f32)
    return np.where(other == 0, g, raw).astype(f32)


def raw_grad_b(grad, a, other):
    with np.errstate(divide="ignore", invalid="ignore"):
        return (-grad * (a * (f32(1) / (other * other)))).astype(f32)


def guard_grad_b(grad, a, other, raw):
    """What the discarded where() at :749 would have written."""
    g = np.where(grad == 0, NAN, np.where(a == 0, NAN, (np.sign(a) * -INF) * np.sign(grad))).astype(f32)
    return np.where(other == 0, g, raw).astype(f32)


def torch_ref(grad, a, other):
    ta = torch.tensor(a, dtype=torch.float64, requires_grad=True)
    tb = torch.tensor(other, dtype=torch.float64, requires_grad=True)
    (ta / tb).backward(torch.tensor(grad, dtype=torch.float64))
    return ta.grad.item(), tb.grad.item()


cases = [(1.0, 2.0, 0.0), (-1.0, 2.0, 0.0), (0.0, 2.0, 0.0), (1.0, 0.0, 0.0),
         (1.0, 2.0, -0.0), (-1.0, 2.0, -0.0), (1.0, -3.0, -0.0)]
print(f"{'grad':>5} {'a':>5} {'other':>6} | {'raw dA':>8} {'guard dA':>9} {'torch dA':>9} | {'raw dB':>8} {'guard dB':>9} {'torch dB':>9}")
for g, a, o in cases:
    g32, a32, o32 = f32(g), f32(a), f32(o)
    ra = raw_grad_a(g32, o32); ga = guard_grad_a(g32, o32, ra)
    rb = raw_grad_b(g32, a32, o32); gb = guard_grad_b(g32, a32, o32, rb)
    ta, tb = torch_ref(g, a, o)
    flag = "" if (np.isnan(ga) and np.isnan(ta)) or ga == ta else "  <-- guard would be WRONG"
    print(f"{g:>5} {a:>5} {o!r:>6} | {ra!s:>8} {ga!s:>9} {ta!s:>9} | {rb!s:>8} {gb!s:>9} {tb!s:>9}{flag}")

print("\nDispatch count (both gradients requested, rounding_mode=None):")
print("  useful: reciprocal, multiply (dA) + square, reciprocal, multiply, neg, multiply (dB) = 7")
print("  discarded guard dA: eqz x2, sign, multiply, where x2                              = 6")
print("  discarded guard dB: eqz x3, sign x2, multiply x2, where x3                        = 10")
print("  -> 16 of 23 device ops allocate a fresh tensor whose result is never read")
```
</details>

Output:

```
 grad     a  other |   raw dA  guard dA  torch dA |   raw dB  guard dB  torch dB
  1.0   2.0    0.0 |      inf       inf       inf |     -inf      -inf      -inf
 -1.0   2.0    0.0 |     -inf      -inf      -inf |      inf       inf       inf
  0.0   2.0    0.0 |      nan       nan       nan |      nan       nan       nan
  1.0   0.0    0.0 |      inf       inf       inf |      nan       nan       nan
  1.0   2.0   -0.0 |     -inf       inf      -inf |     -inf      -inf      -inf  <-- guard would be WRONG
 -1.0   2.0   -0.0 |      inf      -inf       inf |      inf       inf       inf  <-- guard would be WRONG
  1.0  -3.0   -0.0 |     -inf       inf      -inf |      inf       inf       inf  <-- guard would be WRONG
```

## Proposed fix

Delete both `ttnn::where(...)` blocks (and their `eqz`/`sign`/`multiply` operands) from the tensor–tensor `div_bw`, leaving

```cpp
ttnn::multiply(grad_tensor, ttnn::reciprocal(other, output_mem_config), std::nullopt, output_mem_config, input_grad);
```
and
```cpp
ttnn::multiply(ttnn::neg(grad_tensor, ...), ttnn::multiply(input_a, ttnn::reciprocal(ttnn::square(other, ...), ...), ...), std::nullopt, output_mem_config, other_grad);
```
with a comment stating that `reciprocal(±0)` already yields the IEEE-correct signed infinity / NaN, so no explicit zero-divisor branch is needed. 23 launches become 7 with no change in output for `other == +0`, and the `-0.0` behaviour stays correct.

If the intent is instead to keep an explicit guard, it must use `sign(other)`-aware infinities (as the `other_grad` branch already does with `sign(input_a)`), and pass the destination tensor.

## Suggested tests

- `div_bw` with `other` containing `+0.0`, `-0.0`, and `grad == 0` at those positions, compared to `torch.autograd` including the sign of the infinity (`torch.equal` on `torch.sign`, not PCC).
- A dispatch-count assertion (program-cache entries or op-profiler count) for `div_bw(tensor, tensor)` so dead launches cannot return silently.

Source-level analysis; I do not have a card. The dispatch counts are read from the composite and the values from a host model assuming IEEE fp32 `reciprocal`/`square`, which is what the sibling ops rely on as well.

### PR DESCRIPTION
**Pull request by 0xalydev**
### Summary
Resolves #55392.

In `ttnn.div_bw(grad_tensor, input_a, other, ...)` (tensor-tensor overload), two `ttnn::where(...)` zero-divisor guard blocks were evaluated and immediately discarded because the calls omitted the 5th `output` destination argument and their return values were never captured. As a result, 16 out of 23 device op launches (~70%) were allocating temporary tensors whose contents were never used.

Furthermore, IEEE fp32 `reciprocal(+-0.0)` already yields signed infinities and NaNs consistent with PyTorch autograd semantics. Retaining the dead guards with a destination argument would have introduced an erroneous positive-infinity sign overwrite on negative zero (`-0.0`) divisors.

### Changes
- Removed the 16 dead device op dispatches (`ttnn::eqz`, `ttnn::sign`, `ttnn::where`) and unused constants (`t_nan`, `t_inf`, `neg_inf`) in `ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp`.
- Reduced device op launches from 23 down to 7 (70% reduction) when both input gradients are requested, cutting memory overhead and execution latency while preserving exact bit-for-bit IEEE fp32 gradient semantics.

Fixes #55392.

### CODE REVIEW

**Comment by mateusznowakTT** [comment]
@0xalydev : Please include UT that examines the cases covered with this PR.

**Comment by 0xalydev** [comment]
@mateusznowakTT Added unit tests in commit bcc7c9d (	ests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_div.py):

- 	est_bw_binary_div_signed_zeros: tests tensor-tensor div_bw with mixed positive (+0.0) and negative (-0.0) zero divisors against PyTorch autograd golden reference.
- 	est_bw_binary_div_zero_grad_with_signed_zero_divisor: tests div_bw with zero gradients across signed zero divisors.

Both verify IEEE fp32 gradient semantics and parity with PyTorch golden output.
